### PR TITLE
Enable java 8 library desugaring

### PR DESCRIPTION
**Pull request by Vaibhavi1707**
## Purpose / Description
This pull request aims to enable java 8 library desugaring.

## Fixes
Fixes [#7052 ](https://github.com/ankidroid/Anki-Android/issues/7052)

## Approach
The change enables core library desugaring for java 8 libraries.

## How Has This Been Tested?

The changes have been tested on API 22 by adding `import java.time.*`. It is one of the libraries supported by d8 desugaring.  

## Learning (optional, can help others)
I explored desugaring and dex compiler, d8 in Android Studio.
I referred to the following:
https://jakewharton.com/d8-library-desugaring/
https://blog.stylingandroid.com/d8-desugaring/
https://github.com/ankidroid/Anki-Android/issues/7052

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/build.gradle [hunk 1]
```diff
@@ -214,6 +215,8 @@ dependencies {
         resolutionStrategy {
             force 'org.jetbrains:annotations:20.1.0'
         }
+        coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.9' // Setting the version of core desugaring library.
+//    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.4'
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:219]: Please only propose code that is useful, no commented out code should make it into PRs, and while comments are usually a good thing they must add information, this one is redundant I think
```suggestion
        coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.9'
```

File: AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/CardAppearance.java [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import com.ichi2.anki.reviewer.ReviewerCustomFonts;
 import com.ichi2.libanki.Card;
 import com.ichi2.themes.Themes;
+//import java.time.*; This line was added to test the build and run on API 22.
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/CardAppearance.java:8]: this shouldn't be necessary, you can just describe the testing you performed in the PR description
```suggestion
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
Hi there! I like this change, but I think just putting the one line in is sufficient, along with a description of how you tested in the PR itself

**Comment by Vaibhavi1707** [comment]
Hi, I have committed the changes you suggested. I seem to be having some Gradle build issues right now, hence I put this or as a draft. I'll fix it as soon as possible.

**Comment by david-allison-1** [comment]
Just to note: All the checks are passing on CI

**Comment by Vaibhavi1707** [comment]
All the checks have been passed. Opening the PR for review.

**Review by david-allison-1 — APPROVED** [review]
If it passes CI, I'm happy, thanks!

File: AnkiDroid/build.gradle [hunk 3]
```diff
@@ -214,6 +215,7 @@ dependencies {
         resolutionStrategy {
             force 'org.jetbrains:annotations:20.1.0'
         }
+        coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.9'
     }
 
     compileOnly 'org.jetbrains:annotations:20.1.0'
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:221]: Sorry I did not notice this earlier, but you've added it in the wrong spot I think.
We don't want it in the `configurations.all` block - that's a special hack I put in for a specific reason as documented in the comment.

As documented for Android Studio https://developer.android.com/studio/releases#4-0-0-desugar you just add this to the regular dependencies list

```suggestion
    }

    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.9'
    compileOnly 'org.jetbrains:annotations:20.1.0'
```

File: AnkiDroid/build.gradle [hunk 4]
```diff
@@ -154,6 +154,7 @@ android {
         preDexLibraries = preDexEnabled && !ciBuild
     }
     compileOptions {
+        coreLibraryDesugaringEnabled true // Enabling desugaring.
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:157]: As mentioned in my previous review comment, code comments in general are good but: *they must add information* - a comment that does not add information is just more to read + think about that in the end performs no service.

I believe we can all agree that if we are setting a variable called `coreLibraryDesugaringEnabled` to `true` it is obvious we are `// Enabling desugaring.` yes? So the comment adds no value and thus takes time and attention for no purpose.

```suggestion
        coreLibraryDesugaringEnabled true
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
I will admit I am nitpicking on the comment but the point I make there is important for any programmer - only take people's attention in code with important things.

The second bit about where the dependency is, I actually think that's important.

**Comment by Vaibhavi1707** [comment]
I apologize for the inconvenience caused as I'm working with gradle build files and CI for the first time. Thanks for the tip on comments will remember it in future. I have applied your changes and ran it on my end, where I ran into some issues. Hi, I have faced an error in the Gradle build stating the maven is deprecated for gradle 7.0. It suggests to use the maven-publish plugin. On researching further I found that we'll have to replace the uploadArchives task with generatePomFileForArchivesPublication to generate the POM XML

**Comment by mikehardy** [comment]
@Vaibhavi1707 I think what you mention is covered in #5634 - you should post your initial findings there to capture them as they are separate

This change (enabling desugaring) should be very focused and if you could handle the comments I added we could get this change in and release it to make sure it doesn't impact anything with our devs and alpha testers

**Comment by mikehardy** [comment]
I didn't hear back - I just committed the suggestions myself, if they build / pass CI I'll merge them in - would like to see us have the core library desugaring capability

**Comment by Vaibhavi1707** [comment]
Hey, I'm really sorry couldn't get back to this. I had been a bit busy with the past two days. How did we handle the depreciation?

**Comment by mikehardy** [comment]
No worries - open source development means you do your best when you can, and when you don't have time that's okay ;-)

We have not yet - the deprecation is a fully separate issue - already tracked as #5634

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice (one per person for PRs merged in March, regardless of PR quantity) that we try to process OpenCollective payments monthly - it's time for March 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

Thanks!

**Pull request merged**